### ISSUE TITLE
Validation dataset mismatch with image

### ISSUE THREAD

**Issue by wheltz** [ThierryDeruyttere/Talk2Car-Destination#7]
I tried visualizing your validation set and noticed there was a mismatch between command, box and front view and top-down image. I checked talk2car_destination_val.json and found that the file has a total of 1159 items, and the validation dataset has 1162 images.
![image](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/57028913/f6962675-696f-49f6-bd33-fe151874d44a)
In addition, I don't understand why there are many (2 or 3) destinations for a picture in the data set. The original paper says "having three annotators per each command indicate the physical end position of the car after executing the command in the top-down view together with the true intent of the command", but I still don't understand why there are three destinations instead of two, and the several destinations are sometimes far apart and sometimes very close.

**Comment by ThierryDeruyttere** [comment]
Hi,

Thank you for your interest in Talk2Car!
Could you provide me with the index of the sample you are using in the image above?  
I have looked for the exact command in the validation set and I only find that the image with ix 1104 has that specific command in the validation set. Below, I added a screenshot that seems to have the exact same bounding box as in your image but with img_val_1104.jpg. I'm attaching a zip folder to this post that contains the visualization script that I've used. By looking at the dataset it seems like you used the img_val_0.jpg for the 0-th sample in the validation dataset. But in reality the image for the 0-th sample is actually img_val_1104.jpg. Each sample has an 'image' field that indicates the image the command is for. 

![Screenshot from 2023-11-24 14-15-15](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/6353651/d381feef-8127-4616-82fb-44e016a5c13e)

[visualize.zip](https://github.com/ThierryDeruyttere/Talk2Car-Destination/files/13459463/visualize.zip)

Concerning the mismatch between the number of images vs the number of samples in the dataset: thank you for pointing this out. It seems I didn't filter out the images for which there is no command. Good find! The 1162 images come from the original Talk2Car dataset. We filtered some of the images out because their referred object was further than a specific distance from the ego-vehicle (iirc).

For the destination annotations: every annotator indicated one physical end position. The disparity in the distance between these points comes from the annotators themselves. It also reflects the nature of us humans. We all interpret commands in a different way. For a command "stop near object X" people might stop before the object, next to it, or a bit further if they see some parking to safely park instead of stopping in the middle of the road. 

I hope this helps. If you have any other questions, let me know!

**Comment by wheltz** [comment]
Thank you for your reply! 
1. I checked my visualization python file, there should be no bugs, and I checked talk2car_destinationn_val.json. The command "take a left to follow the gray car" is indeed the first command in my json file, and this picture Also the first image of my dataset val_01.
But the top-down image is indeed the same as what you showed. I re-downloaded the front-view image you provided, and found that val_01 is still the same image, so I guess you may have provided the wrong image.zip. You may Need to check it out.
> Concerning the mismatch between the number of images vs the number of samples in the dataset: thank you for pointing this out. It seems I didn't filter out the images for which there is no command. Good find! The 1162 images come from the original Talk2Car dataset. We filtered some of the images out because their referred object was further than a specific distance from the ego-vehicle (iirc).

After this, I sincerely hope that you can provide a data set that excludes images without commands and has a correct front-view image

2. Regarding your explanation of destination annotations, does this mean that there is no best answer among the different explanations for the same command (usually there are three explanations), that is, assuming that there are two destinations, it is speculated that they are relatively close, but they are not the same as the remaining destinations. The distance is very far. We still regard all three explanations as correct, as shown in the figure below.
![image](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/57028913/7734e370-1230-4692-be75-ed6c0106a7c6)

3. In addition, in Figures 3 and 5 in your paper, cyan destination boxes for frontal view and top-down view annotations are provided respectively, as well as heat map visualization of model prediction results in front al view and top-down view. I would be very grateful if you could provide the visualization code for this part.
![image](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/57028913/9b4dbdf9-ef12-403a-81cf-a2c606ba5c13)
![image](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/57028913/699acd59-c9f9-491f-ba18-e6dd8dc69078)

4. Finally, regarding the command_token in the data set, I want to confirm whether it only represents the token sequence of command after bert processing.
Thank you very much, this is a very valuable job

**Comment by ThierryDeruyttere** [comment]
> I checked my visualization python file, there should be no bugs, and I checked talk2car_destinationn_val.json. The command "take a left to follow the gray car" is indeed the first command in my json file, and this picture Also the first image of my dataset val_01. 

The image that belongs to the first command in the validation dataset is val_1104. val_01 on the other hand belongs to the 537th command in the validation set. Maybe I should have been more clear that the ordering in the dataset is not the same ordering for the images! Please have a look at the 'image' key in each dataset sample.

 
![Screenshot from 2023-11-24 17-45-35](https://github.com/ThierryDeruyttere/Talk2Car-Destination/assets/6353651/c05df72b-2741-404c-8a0f-54a82a824784).

Please try out this code (the same as was in the zip file)

```
# This script loads a sample from Talk2Car-Destination and visualizes the following things:
# referred object in top and frontal view
# ego car in top down view
# Destination

import json
from PIL import Image, ImageEnhance
from PIL import ImageDraw
import numpy as np
dataset = "val"
image_ix = 1
#train samples
ds = json.load(open(f"./data/talk2car_destination_{dataset}.json", "r"))

# command = [x for x in ds.values() if x["command"] == "take a left to follow the grey car"]
# Look for sample with frontal_img train_0.jpg
sample = [x for x in ds.values() if x["image"] == f"{dataset}_{image_ix}.jpg"][0]
# sample = sample_annos

# Load images
img = Image.open(f"./data/imgs/img_{sample['image']}").convert("RGB")
top_down = Image.open(f"./data/top_down/{sample['top-down']}").convert("RGB")

drw = ImageDraw.Draw(img)
drw_top = ImageDraw.Draw(top_down)

print(sample["command"])

# Draw on frontal view
(x,y,w,h) = sample['gt_ref_obj_box_frontal']
drw.rectangle([x, y, x+w, y+h], outline="yellow")
img.show()

# Draw on top-down view
flatten = lambda t: [item for sublist in t for item in sublist]

drw_top = ImageDraw.Draw(top_down)
ego_car = np.array(sample["egobbox_top"])
egobbox_polygon = np.concatenate((ego_car, ego_car[0, :][None, :]), 0)

# Draw predictions
for det_obj in sample["all_detections_top"]:
    det_obj = np.array(det_obj)
    det_obj_polygon = np.concatenate((det_obj, det_obj[0, :][None, :]), 0)
    drw_top.polygon(
        flatten(det_obj_polygon.tolist()), fill="#808080", outline="#808080",
    )

ref_obj = np.array(sample["gt_referred_obj_top"])
ref_obj_polygon = np.concatenate((ref_obj, ref_obj[0, :][None, :]), 0)

drw_top.polygon(flatten(egobbox_polygon.tolist()), fill="#ff0000", outline="#ff0000")
drw_top.polygon(
    flatten(ref_obj_polygon.tolist()), fill="yellow", outline="yellow",
)

# Draw annotations on top down
for (x1, y1) in sample["destinations"]:
    color = (255, 0, 255)
    x2 = int(x1 + 3)
    y2 = int(y1 + 3)
    x1 = int(x1 - 3)
    y1 = int(y1 - 3)
    drw_top.ellipse([(x1, y1), (x2, y2)], color)

top_down.show()
```
> After this, I sincerely hope that you can provide a data set that excludes images without commands and has a correct front-view image.

Yes, I'm re-uploading the images and leaving out the excluded images without commands. Thank you for finding this! The front views are, as far as I can see, correct. But again please have a look at the 'image' key in the samples. Apologies if this was unclear.

> Regarding your explanation of destination annotations, does this mean that there is no best answer among the different explanations for the same command (usually there are three explanations), that is, assuming that there are two destinations, it is speculated that they are relatively close, but they are not the same as the remaining destinations. The distance is very far. We still regard all three explanations as correct, as shown in the figure below.

You are indeed correct! We are trying to create a system that can model all *possible* destinations for a given command. That is why we are using a model that outputs a distribution. If the ground truth annotations are all close to each other, then (hopefully) the model learns to output a very dense distribution near those ground truth points. If they are a bit more distant from each other, the model should learn to create a multimodal distribution.

> In addition, in Figures 3 and 5 in your paper, cyan destination boxes for frontal view and top-down view annotations are provided respectively, as well as heat map visualization of model prediction results in frontal view and top-down view. I would be very grateful if you could provide the visualization code for this part.

The visualization code for the top-down is already provided in the test.py file of PDPC, see [here](https://github.com/ThierryDeruyttere/Talk2Car-Destination/blob/60d4eb38d356636245bca42d20836b9e89d0a2f5/baselines/PDPC/test.py#L240-L324).
We also provide the script to project those heatmaps to the frontal view (also in the code above). The cyan bounding boxes visualization is not provided as the models that we have do not output the orientation the car should be in + they output densities. These cyan bounding boxes were just nice to show in the paper. They were also made by taking a screenshot in the annotation tool that we made so we don't even have a visualisation script for them.

> Finally, regarding the command_token in the data set, I want to confirm whether it only represents the token sequence of command after bert processing.

The command token is a token tied to the original talk2car dataset. If you look at the nuscenes dataset (on which Talk2Car is built), they also have unique identifiers (tokens) for each object etc. In the same vein we created unique identifiers (tokens) for our commands. So no they have nothing to do with Bert.

Hope this resolved your questions. In case you have other questions, do let me know!

**Comment by ThierryDeruyttere** [comment]
The new images are online!
